dipy/tracking/utils.py

```diff
@@ -595,6 +596,45 @@ def target(streamlines, target_mask, affine, include=True):
             yield sl
 
 
+@_with_initialize
+def target_line_based(streamlines, target_mask, affine=None, include=True):
+    """Filters streamlines based on whether or not they pass through an ROI,
+    using a line-based algorithm for compressed tract.
+
+    Parameters
+    ----------
+    streamlines : iterable
+        A sequence of streamlines. Each streamline should be a (N, 3) array,
+        where N is the length of the streamline.
+    target_mask : array-like
+        A mask used as a target. Non-zero values are considered to be within
+        the target region.
+    affine : array (4, 4)
+        The affine transform from voxel indices to streamline points.
+    include : bool, default True
+        If True, streamlines passing through `target_mask` are kept. If False,
+        the streamlines not passing through `target_mask` are kept.
+
+    Returns
+    -------
+    streamlines : generator
+        A sequence of streamlines that pass through `target_mask`.
+
+    See Also
+    --------
+    density_map
+    """
+    target_mask = np.array(target_mask, dtype=np.uint8, copy=True)
+    lin_T, offset = _mapping_to_voxel(affine, voxel_size=None)
+    yield
```

Comment:
Since you use the ```@_with_initialize``` decorator, it will be initialized up to here. However, on the first real iteration on this generator, the whole filtering code will be executed. I guess that's a matter of philosophy, but wouldn't it make more sense to call ```streamlines_in_mask()``` directly in the initialization, and then yield in constant time for all the iterations?